verifiable-db/src/row_tree/leaf.rs

```diff
@@ -20,31 +20,31 @@ use recursion_framework::{
 };
 use serde::{Deserialize, Serialize};
 
-use crate::cells_tree;
+use crate::cells_tree::{self, Cell, CellWire};
 
-use super::{public_inputs::PublicInputs, IndexTuple, IndexTupleWire};
-use derive_more::{Constructor, Deref, From};
+use super::public_inputs::PublicInputs;
 
 // new type to implement the circuit logic on each differently
 // deref to access directly the same members - read only so it's ok
-#[derive(Clone, Debug, Deref, From, Constructor)]
-pub struct LeafCircuit(IndexTuple);
+#[derive(Clone, Debug, From, Into)]
+pub struct LeafCircuit(Cell);
 
-#[derive(Clone, Serialize, Deserialize, Deref, From)]
-pub(crate) struct LeafWires(IndexTupleWire);
+#[derive(Clone, Serialize, Deserialize, From, Into)]
+pub(crate) struct LeafWires(CellWire);
 
 impl LeafCircuit {
     pub(crate) fn build(b: &mut CircuitBuilder<F, D>, cells_pis: &[Target]) -> LeafWires {
         let cells_pis = cells_tree::PublicInputs::from_slice(cells_pis);
         // D(index_id||pack_u32(index_value)
-        let tuple = IndexTupleWire::new(b);
-        let d1 = tuple.digest(b);
-        // D(proof.DC + D(index_id||pack_u32(index_value)))
-        let cells_digest = cells_pis.digest_target();
-        let input_digest = b.curve_add(cells_digest, d1);
-        let row_digest = b
-            .map_to_curve_point(&input_digest.to_targets())
-            .to_targets();
+        let tuple = CellWire::new(b);
+        // set the right digest depending on the multiplier and accumulate the ones from the public
+        // inputs of the cell root proof
+        let split_digest = tuple.split_and_accumulate_digest(b, cells_pis.split_digest_target());
+        // final_digest = HashToInt(mul_digest) * D(ind_digest)
+        // NOTE This additional digest is necessary since the individual digest is supposed to be a
+        // full row, that is how it is extracted from MPT
+        let final_digest = split_digest.cond_combine_to_row_digest(b);
```

Comment:
In the row circuits I think that instead of having to check whether `multiplier ==NEUTRAL`, it could be a bit cheaper to have an input flag that tells whether there is not a multiplier to be added. We can call this input flag something like `is_combined_table`, so that its semantic is clear. This will allow to remove the need to have `curve_eq` operation, which might require a couple of additional rows. Though I see it's not a dramatic improvement, so if you think it would make the APIs too complex we could also avoid it.

Reply:
Mhhh it doesn't make the API complex if i still rely on  the check `point == neutral` but outside of circuit, wdyt ?

Reply:
Yeah, that would make sense. Though I suppose you should also take into account the `multiplier` flag for the current row? Like, if `point == neutral` for all the other cells but `multiplier = true` for the secondary index, then the scalar multiplication would still need to be performed, right?